[ConstraintSystem] Bind Self to correct contextual type for nested types in protocol

xedin opened the pull request:
If something that we are trying to contextually bind is a nested type inside
protocol or protocol extension, let's try to find the innermost conforming type
from the current declaration context and map Self parameter of the protocol
to that nominal type. Since nested types in protocols aren't yet implemented this
is going to result in failure, but that's better than crashing.

Resolves: rdar://problem/36449760

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

DougGregor commented on lib/AST/NameLookup.cpp:
```diff
@@ -1646,9 +1646,18 @@ bool DeclContext::lookupQualified(Type type,
+      // we have a nested type in generic context.
+      if (DC->getAsProtocolOrProtocolExtensionContext() ||
+          (DC->isLocalContext() && DC->isGenericContext()))
+        return false;
```
It seems like overkill to completely bail out on returning any results from anything. Should we just "continue" here?

xedin replied:
Even if we continue this is not going to produce any results because this if else sequence would end up with empty `stack` which I think is the idea since we don't really want to allow any lookup on the types like that, no?

xedin replied:
Ok, I think I understand what you mean, will push the changes soon...

slavapestov commented on validation-test/compiler_crashers_2_fixed/0138-rdar36449760.swift:
```diff
@@ -0,0 +1,8 @@
+// RUN: not %target-swift-frontend %s -typecheck
+
+class C {
```
I don't see how this test case demonstrates a problem, since C does not conform to protocol A.

xedin replied:
I'm about to add 2 more cases here which might be better - `foo() -> A.B` and `foo(b: B) { _ = b.answer() }`

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -1646,9 +1646,18 @@ bool DeclContext::lookupQualified(Type type,
+      // Refuse to return results for situation when
+      // we have a nested type in generic context.
+      if (DC->getAsProtocolOrProtocolExtensionContext() ||
+          (DC->isLocalContext() && DC->isGenericContext()))
```
In the past we've had a lot of crashes caused by "bail out" hacks like this. Can you fix the underlying issue instead of hacking around it? If you hack around it, you will probably just find a similar crash later. Eventually we might support types nested inside protocols, and it would be a pain to hunt down all the hacks put in to work around the underlying problems.

xedin replied:
As far as I understand to actually fix the problem would mean to implement support for nested types in generics because the crash happens in `openGeneric`'s `bindArchetypesFromContext` when it tries to access generic environment of the declaration context.

slavapestov replied:
We already support types nested inside generic types, just not protocols or local contexts. It shouldn't be hard to make openGeneric() do something self-consistent here. In fact I put in bindArchetypesFromContext() to fix crashes with types inside local contexts (even though the rest of the compiler doesn't support it yet). So fixing up that method would be preferable to putting in a hack that makes it unreachable altogether. If it becomes unreachable it will just bitrot and we'll have more debugging to catch up on when the rest of the compiler learns about types in local contexts.

xedin replied:
Ok, what I was going for originally was fixing `bindArchetypesFromContext` but I talked to @DougGregor  and it seemed like making lookup not return results in such case would be good idea. I'll do some more digging.

slavapestov commented on lib/AST/NameLookup.cpp:
```diff
@@ -1696,7 +1696,18 @@ bool DeclContext::lookupQualified(Type type,
+        // Refuse to return results for situation when
+        // we have a nested type in generic context.
+        if (DC->getAsProtocolOrProtocolExtensionContext() ||
+            (DC->isLocalContext() && DC->isGenericContext()))
```
openGeneric() should already support generic types inside local contet via bindArchetypesFromContext(), and eventually I want the rest of the compiler to support it too. Please don't make this change.

xedin committed:
[ConstraintSystem] Bind Self to correct contextual type for nested types in protocol

If something that we are trying to contextually bind is a nested type inside
protocol or protocol extension, let's try to find the innermost conforming type
from the current declaration context and map Self parameter of the protocol
to that nominal type. Since nested types in protocols aren't yet implemented this
is going to result in failure, but that's better than crashing.

Resolves: rdar://problem/36449760

xedin commented:
@slavapestov @DougGregor I've changed it so we actually bind Self parameter to the proper type in `bindArchetypesFromContext` and fixed the test to declare conformance to A. WDYT?

xedin commented:
I'm going to change the message etc. later on if this is the right approach to fixing this.

DougGregor commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1015,8 +1015,42 @@ static void bindArchetypesFromContext(
+          } while ((DC = DC->getParent()));
+
+          assert(conformingEnclosingDC);
+          contextTy = conformingEnclosingDC->getDeclairedTypeInContext();
```
Typo `getDeclairedTypeInContext`?

xedin replied:
Yes, sorry, I must have accidentally added "i" :(

DougGregor commented on validation-test/compiler_crashers_2_fixed/0138-rdar36449760.swift:
```diff
@@ -0,0 +1,15 @@
+  func foo() -> B {}
+  func bar() -> A.B {}
+  func baz(b: B) {
+    _ = b.answer
```
It'd be good to have these same tests nested (1) in a generic type, and (2) nested within yet another type inside C that does not conform to A.

xedin replied:
Sounds good, will do!

DougGregor commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1015,8 +1015,42 @@ static void bindArchetypesFromContext(
+        if (genericEnv) {
+          contextTy = genericEnv->mapTypeIntoContext(paramTy);
+        } else {
+          auto &TC = cs.TC;
```
The general approach looks good, but please pull this computation out into a separate function.

xedin replied:
I'm going to make `getInnermostConformingDC` which returns `NominalTypeDecl *`.

xedin commented:
@DougGregor All done, WDYT?

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

DougGregor reviewed (approved):
LGTM, thank you!
